content/blog/2022/2022-08-25-schedule-cron-job-in-node.md

```diff
@@ -0,0 +1,318 @@
+- In the `Signing in to Google` section select `App password` option.
+- If the `App password` option is unavailable, set up 2-Step Verification for your account.
+- Select the app (mail) and the current device you want to generate the app password for.
+- Click Generate.
+- Copy the generated 16-character code in the yellow bar on your device.
+
+To create our email scheduler application Insert the following code into the `index.js` file:
+```javascript
+const express = require("express");
+const cron = require("node-cron");
+const nodemailer = require("nodemailer");
+app = express();
+
+//Send email after 1 minute
+cron.schedule("1 * * * *", function () {
+  mailService();
+});
+
+function mailService() {
+  let mailTransporter = nodemailer.createTransport({
+    service: "gmail",
+    auth: {
+      user: "<your-email>@gmail.com",
+// use generated app password for gmail
+      pass: "***********",
+    },
+  });
+
+  // Setting credentials
+  let mailDetails = {
+    from: "<your-email>@gmail.com",
+    to: "<user-email>@gmail.com",
+    subject: "Test Mail using Cron Job",
+    text: "Node.js Cron Job Email Demo Test from Reflectoring Blog",
+  };
+
+  // Sending Email
+  mailTransporter.sendMail(mailDetails, function (err, data) {
+    if (err) {
+      console.log("Error Occured", err.message);
+    } else {
+      console.log("---------------------");
+      console.log("Email sent successfully");
+    }
+  });
+}
+
+app.listen(3000, () => {
+  console.log("Application Listening.....");
+});
+
+```
+In the above code we are using the node-mailer and node-cron dependencies we installed earlier. The `node-mailer` dependency gives us the feature of sending e-mails from our Node.js applications, using any email service of your choice.
+
+We scheduled our mail to be sent at `At 1 minute` interval using Gmail.
+
+Run the script `node index.js`
+
+`output:`
+```
+Application Listening.....
+---------------------
+Email Sent Successfully
+```
+Check your inbox to confirm the email is sent.
+
+### 3. Writing to a Log File
+Cron jobs can be used to schedule logging tasks in a system. Let's say something happened, like a network delay or a warning message. To track server status, we can keep a log for a specific period of time.
+
+In this example, we will be making use of some `node.js` inbuilt modules to get information about the current processes or runtime of our `node.js` program:
+
+Insert this code in the `index.js` file to generate the application log file at our scheduled time:
+
+```javascript
+const process = require("process");
+const fs = require("fs");
+const cron = require("node-cron");
+const express = require("express");
+
+app = express();
+
+// Setting a cron job for every 15 seconds
+cron.schedule("*/15 * * * * *", function () {
+  let heap = process.memoryUsage().heapUsed / 1000000;
+
+  let data = `Memory usage by heapUsed ${heap}MB
+  ${new Date().toUTCString()}
```

Comment:
It makes more sense to format this log message in CSV format to make it more machine-readable: `2022-08-30 10:00:23, 8234265, 35.43%`

Where the first column is the date/time, the second column are the bytes of memory used, and the third column is the cpu used (can you add that to the code?).

This would make the value of this use case really shine: automatic monitoring over time. You could then mention that the generated CSV file could be imported into a spreadsheet to create graphs.